synfig-core/src/synfig/os.cpp

```diff
@@ -0,0 +1,421 @@
+**
+**	You should have received a copy of the GNU General Public License
+**	along with Synfig.  If not, see <https://www.gnu.org/licenses/>.
+**	\endlegal
+*/
+/* ========================================================================= */
+
+/* === H E A D E R S ======================================================= */
+
+#ifdef USING_PCH
+#	include "pch.h"
+#else
+#ifdef HAVE_CONFIG_H
+#	include <config.h>
+#endif
+
+#include "os.h"
+
+#if defined(HAVE_FORK) && defined(HAVE_PIPE) && defined(HAVE_WAITPID)
+ #define UNIX_PIPE_TO_PROCESSES
+ #include <cstdlib> // for system()
+ #include <fcntl.h> // for O_ flags
+ #include <sys/wait.h> // for waitpid()
+ #include <unistd.h> // for popen()
+#elif defined(_WIN32)
+ #define WIN32_PIPE_TO_PROCESSES
+ #define WIN32_LEAN_AND_MEAN
+ #include <windows.h>
+ #include <shellapi.h> // for ShellExecuteW
+#else
+	#error There are no known APIs for creating child processes
+#endif
+
+#include <ETL/stringf>
+#include <synfig/general.h>
+#include <synfig/localization.h>
+
+#endif
+
+/* === U S I N G =========================================================== */
+
+using namespace synfig;
+
+/* === P R O C E D U R E S ================================================= */
+
+void OS::RunArgs::push(const std::vector<std::string> &args)
+{
+	run_args.insert(run_args.end(), args.begin(), args.end());
+}
+
+void OS::RunArgs::push(const std::vector<const char *> &args)
+{
+	run_args.insert(run_args.end(), args.begin(), args.end());
+}
+
+void OS::RunArgs::push(const std::string &item)
+{
+	run_args.push_back(item);
+}
+
+void OS::RunArgs::push(std::string &&item)
+{
+	run_args.emplace_back(std::move(item));
+}
+
+void OS::RunArgs::push(const char *__format, ...)
+{
+	va_list args;
+	va_start(args,__format);
+	run_args.emplace_back(etl::vstrprintf(__format, args));
+	va_end(args);
+}
+
+void OS::RunArgs::push_pair(const std::string &key, const std::string &value)
+{
+	run_args.push_back(key);
+	run_args.push_back(value);
+}
+
+void OS::RunArgs::push_pair_filename(const std::string &key, const std::string &filename)
+{
+	run_args.push_back(key);
+	push_filename(filename);
+}
+
+void OS::RunArgs::push_filename(const std::string &filename)
+{
+#ifdef WIN32_PIPE_TO_PROCESSES
+	run_args.push_back('"'+filename+'"');
+#else
+	run_args.push_back(filename);
+#endif
+}
+
+void OS::RunArgs::push_filename(std::string &&filename)
+{
+#ifdef WIN32_PIPE_TO_PROCESSES
+	run_args.emplace_back(std::move('"'+filename+'"'));
+#else
+	run_args.emplace_back(std::move(filename));
+#endif
+}
+
+void OS::RunArgs::push_filename(const char *filename)
+{
+#ifdef WIN32_PIPE_TO_PROCESSES
+	run_args.emplace_back(etl::strprintf("\"%s\"", filename));
+#else
+	run_args.emplace_back(filename);
+#endif
+}
+
+void OS::RunArgs::clear()
+{
+	run_args.clear();
+}
+
+std::string OS::RunArgs::get_string() const
+{
+	std::string ret;
+	for (const auto& arg : run_args) {
+		ret.append(arg + " ");
+	}
+	ret.pop_back();
```

Comment:
The behavior is undefined if the string is empty. (https://en.cppreference.com/w/cpp/string/basic_string/pop_back)